### PR TITLE
8214976: Warn about uses of functions replaced for portability

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this new mechanism for "poisoning" uses of certain functions,
along with marking specific uses as still being permitted.  One use for this
is to discourage use of external library functions for which HotSpot provides
extended versions or portability shims.  Another is to prevent uses of certain
functions considered "insecure", promoting alternatives or requiring strong
scrutiny.

The new mechanism is defined in compilerWarnings.hpp, with compiler-specific
implementations.  Comments in the code describe the details.  For some
compilers we didn't find any features that would permit such poisoning.  As a
result, only shared code has complete coverage.

For gcc and clang we use an attribute that warns about references to an
annotated function, and scoped suppression of that warning.  This
functionality is provided since gcc9 and clang14.  (I haven't tested with
clang; version 14 is more recent than I have access for.  But the docs
describe an explicit match to the gcc functionality.)

For Visual Studio I was unable to find a workable poisoning mechanism.  Using
deprecation annotations and scoped suppression of the consequent warnings
seemed plausible, but ran into problems.  (See comments in code for details.)
But we provide scoped suppression of deprecation warnings anyway.  Many of the
functions we'd poison are already conditionally deprecated under the control
of macros like _CRT_SECURE_NO_WARNINGS and _CRT_NONSTDC_NO_DEPRECATE.  Once
we've removed or annotated all calls associated with one of those macros we
can remove the macro from our builds, preventing new uses from creeping into
unshared code.

The aix-ppc port doesn't provide poisoning support.  The version of clang used
by xlclang is older than the warning attribute and suppression.

Along with the new mechanisms, a few functions are poisoned by this change.
More functions can be added as folks have time to address their uses.

The poisoning declarations are in globalDefinitions.hpp, since the signatures
may use types that have platform-specific definitions or locations that are
provided within that header.  That header is included (directly or indirectly)
nearly everywhere, so provides good coverage.

The rest of the changes are to address existing uses of the newly forbidden
functions, either using something else or annotating them as permitted uses.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8214976](https://bugs.openjdk.org/browse/JDK-8214976): Warn about uses of functions replaced for portability

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [b9cd6f3f](https://git.openjdk.java.net/jdk/pull/8982/files/b9cd6f3ff1d65883862cde917ce40817fbfe82ac)
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [b9cd6f3f](https://git.openjdk.java.net/jdk/pull/8982/files/b9cd6f3ff1d65883862cde917ce40817fbfe82ac)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [b9cd6f3f](https://git.openjdk.java.net/jdk/pull/8982/files/b9cd6f3ff1d65883862cde917ce40817fbfe82ac)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8982/head:pull/8982` \
`$ git checkout pull/8982`

Update a local copy of the PR: \
`$ git checkout pull/8982` \
`$ git pull https://git.openjdk.java.net/jdk pull/8982/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8982`

View PR using the GUI difftool: \
`$ git pr show -t 8982`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8982.diff">https://git.openjdk.java.net/jdk/pull/8982.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
Hi Kim,
This seems fine to me.
Thanks.

**Review by tschatzl — APPROVED** [review]

**Review by coleenp — APPROVED** [review]
Looks good! thank you for this change.

File: test/hotspot/gtest/unittest.hpp [hunk 1]
```diff
@@ -94,7 +97,7 @@
       }                                                             \
     }                                                               \
     fprintf(stderr, "OKIDOKI");                                     \
-    exit(0);                                                        \
+    gtest_exit_from_child_vm(0);                                               \
```
[the remark this answers is not in the document]
Comment (kimbarrett) [reply test/hotspot/gtest/unittest.hpp:100]: Fixed.

**Comment by kimbarrett** [comment]
Thanks for reviews @dholmes-ora , @tschatzl , and @coleenp .

**Pull request closed without merging**